### PR TITLE
Adding API for notification rules in loadtesting

### LINKED ISSUE #31079: [Azure Load Testing - Azure Load Testing] API Review
New API Review meeting has been requested.

**Service Name**: Azure Load Testing - Azure Load Testing
**Review Created By**: Radhika Jain
**Review Date**: 11/07/2024 08:00 AM PT
**Release Plan**: [1335](https://web.powerapps.com/apps/821ab569-ae60-420d-8264-d7b5d5ef734c?release-plan-id=9dd7d3ff-3d2e-ef11-840a-6045bd015e50)
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/31078
**Hero Scenarios Link**: [here](https://microsoftapc.sharepoint.com/:w:/t/IDC-CloudNativeTesting/EW9AilLjW6NNr-UKeJN1m7IBsuxAazBf0HZRk9-ooHbrYg?e=w1teJb)
**Core Concepts Doc Link**: Not Provided

**Description**: 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=d85ef282-608c-ef11-ac21-00224806dc21)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by radhikajain99**
# Data Plane API Specification Update Pull Request

Added new API for `notification-rules` to the service to enable Notifications for various events in Azure Load Testing. This will allow the user to create a `NotificationRule` where they will specify when and how to receive notifications. They can specify the `scope` at which they want to receive notifications. Currently we will support `Tests` scope which can be expanded in the future to support more scopes like `Resource`, `Trigger`, etc. Based on the scope selected, the user can provide further filtering. For `Tests` scope, the user can provide a list of `testIds` for which they want to receive notifications. The user will also provide a list of `events` they want to be notified for. It can include various events like `TestRunEnded`, `TestRunStarted`, `TriggerEnded`, etc. and the user can further filter by specifying `condition` inside the event. They will provide a list of `actionGroupIds` which will have the action group resource id which will be used to send the notification.

This API is added to an existing preview version `2024-12-01-preview` since we have not released that preview version yet and there are no changes in any of the existing paths that version.

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## API Info: The Basics

Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc: [API version 2024-07-01-preview](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/loadtestservice/data-plane/Microsoft.LoadTestService/preview/2024-07-01-preview/loadtestservice.json)
* Updated paths: `notification-rules`

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories) 
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

fix https://github.com/Azure/azure-rest-api-specs/issues/31079

### CODE REVIEW

File: specification/loadtestservice/LoadTestService/models.notificationrule.tsp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  @maxLength(50)
+  @minLength(2)
+  @doc("The name of the notification rule.")
+  displayName: string;
+
+  @doc("The action groups to notify.")
+  actionGroupIds: string[];
+
+  @doc("The scope of the notification rule.")
+  scope: NotificationScopeType;
```
Comment (radhikajain99) [inline specification/loadtestservice/LoadTestService/models.notificationrule.tsp:124]: Use kind?

File: specification/loadtestservice/LoadTestService/models.notificationrule.tsp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+@added(APIVersions.v2024_11_01_preview)
+@doc("Tests Notification rule model.")
+model TestsNotificationRule extends NotificationRule {
+  @doc("Scope of type Tests.")
+  scope: NotificationScopeType.Tests;
+
+  @doc("The test ids to include. If not provided, the notification will be sent for all testIds.")
+  testIds?: string[];
+
+  @doc("The events to send notifications for.")
```
[the remark this answers is not in the document]
Comment (radhikajain99) [reply specification/loadtestservice/LoadTestService/models.notificationrule.tsp:138]: Specify key is identifier for the event

File: specification/loadtestservice/LoadTestService/models.notificationrule.tsp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+@added(APIVersions.v2024_11_01_preview)
+@doc("TestRunEnded Event condition.")
+model TestRunEndedEventCondition {
+  @doc("The test run statuses to send notification for.")
+  testRunStatuses?: Status[];
+
+  @doc("The test run results to send notification for.")
+  testRunResults?: PFTestResult[];
```
Comment (radhikajain99) [inline specification/loadtestservice/LoadTestService/models.notificationrule.tsp:89]: Check if this is needed

File: specification/loadtestservice/LoadTestService/models.notificationrule.tsp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  @doc("Event type for trigger ended event.")
+  name: NotificationEventType.TriggerEnded;
+}
+
+@added(APIVersions.v2024_11_01_preview)
+@doc("The notification event when the event type is TriggerDisabled.")
+model TriggerDisabledNotificationEvent extends TestsNotificationEvent {
+  @doc("Event type for trigger disabled event.")
+  name: NotificationEventType.TriggerDisabled;
+}
```
Comment (radhikajain99) [inline specification/loadtestservice/LoadTestService/models.notificationrule.tsp:80]: Check if Trigger events should be added to a separate Trigger scope

File: specification/loadtestservice/LoadTestService/models.notificationrule.tsp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+@doc("Tests Notification rule model.")
+model TestsNotificationRule extends NotificationRule {
+  @doc("Scope of type Tests.")
+  scope: NotificationScopeType.Tests;
+
+  @doc("The test ids to include. If not provided, the notification will be sent for all testIds.")
+  testIds?: string[];
+
+  @doc("The events to send notifications for.")
+  events: Record<TestsNotificationEvent>;
```
Comment (radhikajain99) [inline specification/loadtestservice/LoadTestService/models.notificationrule.tsp:139]: Rename to eventFilters

File: specification/loadtestservice/data-plane/Microsoft.LoadTestService/preview/2024-11-01-preview/examples/CreateOrUpdateNotificationRule.json [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        "displayName": "Sample_Notification_Rule",
+        "actionGroupIds": [
+          "/subscriptions/12345678-1234-1234-1234-123456789012/resourceGroups/sampleRG/providers/microsoft.insights/actionGroups/sampleActionGroup"
+        ],
+        "scope": "Tests",
+        "testIds": [
+          "12345678-1234-1234-1234-123456789012"
+        ],
+        "eventFilters": {
+          "event1": {
```
Comment (prativen) [inline specification/loadtestservice/data-plane/Microsoft.LoadTestService/preview/2024-11-01-preview/examples/CreateOrUpdateNotificationRule.json:65]: nit: might want to change these to include filter in the name as well (though I agree this is user generated string)

Comment (radhikajain99) [reply specification/loadtestservice/data-plane/Microsoft.LoadTestService/preview/2024-11-01-preview/examples/CreateOrUpdateNotificationRule.json]: Done

File: specification/loadtestservice/LoadTestService/models.notificationrule.tsp (hunk 1 above)
Comment (radhikajain99) [reply specification/loadtestservice/LoadTestService/models.notificationrule.tsp:124]: Prefer scope as per internal discussion

**Review by mitsha-microsoft — APPROVED** [review]

**Pull request merged**